### PR TITLE
Pootle formats

### PR DESCRIPTION
**Pull request by phlax**
Adds models and plugin setup for format registry, which can be used by `Project` and `Store` in the case of having mutiple filetype formats - see #4929

### CODE REVIEW

File: pootle/apps/pootle_app/forms.py [hunk 1]
```diff
@@ -57,28 +56,28 @@ def __init__(self, *args, **kwargs):
...
+        value = self.cleaned_data.get('filetypes', [])
+        if not self.instance.pk:
+            return value
+        for filetype in self.instance.filetypes.all():
+            if filetype not in value:
+                has_stores = Store.objects.filter(
+                    translation_project__project=self.instance, filetype=filetype)
+                if has_stores.exists():
+                    raise forms.ValidationError(
+                        "You cannot remove a filetype ('%s') from a Project, "
```
Comment (unho) [inline pootle/apps/pootle_app/forms.py]: I think this should be marked for translation.

File: pootle/static/js/admin/components/Project/ProjectForm.js [hunk 2]
```diff
@@ -78,13 +78,14 @@ const ProjectForm = React.createClass({
           />
           <FormElement
             type="select"
+            multiple
             clearable={false}
-            options={model.getFieldChoices('localfiletype')}
-            label={gettext('File Type')}
+            options={model.getFieldChoices('filetypes')}
+            label={gettext('Filetypes')}
```
Comment (unho) [inline pootle/static/js/admin/components/Project/ProjectForm.js]: I would keep the whitespace.

File: pootle/static/js/shared/models/project.js [hunk 3]
```diff
@@ -14,16 +14,18 @@ import AdminAPIMixin from 'mixins/admin_api';
...
+      code: '',
+      fullname: '',
+      checkstyle: 'standard',
+      treestyle: 'auto',
+      source_language: '',
+      ignoredfiles: '',
+      filetypes: [],
+      screenshot_search_prefix: '',
+      disabled: false,
+    };
```
Comment (unho) [inline pootle/static/js/shared/models/project.js]: Why are there so many changed lines here? The previous syntax didn't work?

File: pootle/apps/pootle_app/views/admin/projects.py [hunk 4]
```diff
@@ -30,11 +30,16 @@ def get_context_data(self, **kwargs):
         except Language.DoesNotExist:
             default_language = languages[0].id
 
+        filetypes = []
+        for info in formats.get().values():
+            filetypes.append(
+                [info["pk"], "%s(%s)" % (info["title"], info["extension"])])
```
Comment (unho) [inline pootle/apps/pootle_app/views/admin/projects.py]: No space between the title and the parentheses? I wonder if this string could be provided by the format itself.

File: pootle/apps/pootle_app/project_tree.py [hunk 5]
```diff
@@ -49,10 +50,11 @@ def match_template_filename(project, filename):
     :param:`project`.
     """
     name, ext = os.path.splitext(os.path.basename(filename))
+    project_filetypes = ProjectFiletypes(project)
 
     # FIXME: is the test for matching extension redundant?
-    if ext == os.path.extsep + project.get_template_filetype():
-        if ext != os.path.extsep + project.localfiletype:
+    if ext in project_filetypes.template_extensions:
```
Comment (unho) [inline pootle/apps/pootle_app/project_tree.py]: `ext` here is for example `.pot` but I think `project_filetypes.template_extensions` is `.pot`. It won't match.

File: pootle/static/js/shared/models/project.js (hunk 3 above)
Comment (phlax) [inline pootle/static/js/shared/models/project.js]: im not sure its necessary actually - i had some issue around passing arrays

File: pootle/apps/pootle_app/project_tree.py (hunk 5 above)
Comment (phlax) [inline pootle/apps/pootle_app/project_tree.py]: no - `project_filetypes.template_extensions` returns a list see `pootle_format.utils`

Comment (unho) [inline pootle/apps/pootle_app/project_tree.py]: Curiously several lines below you did the right thing.

File: pootle/apps/pootle_app/project_tree.py [hunk 6]
```diff
@@ -188,9 +196,12 @@ def create_or_resurrect_store(file, parent, name, translation_project):
...
                   path=store.pootle_path, store=store.id)
     except Store.DoesNotExist:
-        store = Store(file=file, parent=parent,
-                      name=name, translation_project=translation_project)
-
+        store = Store.objects.create(
+            file=file, parent=parent,
+            name=name, translation_project=translation_project)
+        store.filetype = ProjectFiletypes(
+            translation_project.project).choose_filetype(store.name)
```
Comment (unho) [inline pootle/apps/pootle_app/project_tree.py]: Is it not better to move this to `Store` model?

File: pootle/apps/pootle_app/project_tree.py (hunk 5 above)
Comment (unho) [inline pootle/apps/pootle_app/project_tree.py]: Sure, I misstyped and it returns a list. But the list is `[".pot"]` while `ext` is `".pot"`.

Comment (unho) [inline pootle/apps/pootle_app/project_tree.py]: Note the missing dot in what `project_filetypes.template_extensions` returns.

File: pootle/static/js/admin/components/Project/ProjectForm.js (hunk 2 above)
Comment (phlax) [inline pootle/static/js/admin/components/Project/ProjectForm.js]: no - its called filetype throughout the system, this makes it consistent

File: pootle/apps/pootle_app/project_tree.py (hunk 6 above)
Comment (phlax) [inline pootle/apps/pootle_app/project_tree.py]: i think this logic is actually in the manager - so probs isnt necessary - ill check

File: pootle/static/js/admin/components/Project/ProjectForm.js (hunk 2 above)
Comment (julen) [inline pootle/static/js/admin/components/Project/ProjectForm.js]: Not gonna bikeshed, but as a localizer I don't recall having to translate _filetype_, as opposed to _file type_. Feel free to check any online glossaries available out there (Mozilla, Microsoft...).

File: pootle/apps/pootle_app/project_tree.py (hunk 5 above)
Comment (phlax) [inline pootle/apps/pootle_app/project_tree.py]: k, i understand now, good spot, thanks @unho

Comment (phlax) [inline pootle/apps/pootle_app/project_tree.py]: fixed and test added

File: pootle/apps/pootle_app/forms.py [hunk 7]
```diff
@@ -57,28 +56,28 @@ def __init__(self, *args, **kwargs):
...
+        if not self.instance.pk:
+            return value
+        for filetype in self.instance.filetypes.all():
+            if filetype not in value:
+                has_stores = Store.objects.filter(
+                    translation_project__project=self.instance, filetype=filetype)
+                if has_stores.exists():
+                    raise forms.ValidationError(
+                        _("You cannot remove a filetype ('%s') from a Project, "
+                          "if there are Stores of that filetype"
```
Comment (unho) [inline pootle/apps/pootle_app/forms.py]: Use "file type" rather than "filetype".

**Comment by unho** [comment]
lgtm

**Comment by ta2-1** [comment]
@phlax thanks for your great work. lgtm.

**Pull request merged**